cmd/entrypoint/syntheticauth.go

```diff
@@ -40,13 +43,70 @@ func ReconcileAuthServices(ctx context.Context, sh *SnapshotHolder, deltas *[]*k
 	syntheticAuthExists := false
 	for _, authService := range sh.k8sSnapshot.AuthServices {
 		// Keep any AuthServices already using protocol_version: v3
-		if authService.ObjectMeta.Name == "synthetic-edge-stack-auth" {
-			syntheticAuthExists = true
-		} else if authService.Spec.ProtocolVersion == "v3" {
-			authServices = append(authServices, authService)
+		if authService.Spec.ProtocolVersion == "v3" {
+			injectSyntheticAuth = false
+			if authService.ObjectMeta.Name == "synthetic-edge-stack-auth" {
+				syntheticAuthExists = true
+			} else {
+				authServices = append(authServices, authService)
+			}
 		}
 	}
-	if len(authServices) == 0 && !syntheticAuthExists {
+
+	// Also loop over the annotations and remove authservices that are not v3. We do
+	// this by looping over each entry in the annotations map, removing all the non-v3
+	// AuthService entries, and then removing any keys that end up with empty lists.
+	//
+	// keysToDelete tracks keys to delete, since I never trust deleting things from
+	// collections while iterating over them.
+	keysToDelete := make([]string, 0)
+
+	// OK. Loop over all the keys and their corrauthServicesesponding lists of annotations...
+	for key, list := range sh.k8sSnapshot.Annotations {
+		// ...and build up our edited list of things.
+		editedList := snapshot.AnnotationList{}
+
+		for _, obj := range list {
+			switch annotationObj := obj.(type) {
+			case *v3alpha1.AuthService:
+				// This _is_ an AuthService, so we'll check its protocol version.
+				// Anything other than v3 gets tossed.
+				//
+				// Note also that AuthServices other than getambassador.io/v3 cannot
+				// have a protocol_version, but whatever -- that'll be something not
+				// equal to v3, so it'll get tossed, and that's what we want.
+				if annotationObj.Spec.ProtocolVersion == "v3" {
+					// Whoa, it's a v3! Keep it.
+					editedList = append(editedList, annotationObj)
+					authServices = append(authServices, annotationObj)
+					injectSyntheticAuth = false
+				}
+			default:
+				// This isn't an AuthService at all, so we'll keep it.
+				//
+				// XXX There's optimization to do here: if there are no AuthServices
```

Comment:
The only optimization I see here is that we could check if the list contains any AuthServices at all before allocating the memory to copy all the annotations. This is probably preferable to copying it all every time in cases where a user has a ton of annotation config since iterating over a few thousand items is fast, but loading a copy of all those items uses memory if we dont need to in the event that there are no AuthServices.

With that change...
Best case: There are a ton of Annotations, we iterate over them and find there are no AuthServices, then dont do any copying or editing. 

Worst case: There are a ton of Annotations, we iterate over them and find there is an AuthService we need to drop, but we have to iterate over n-1 items in the list of annotations to find it, then we iterate over it again but this time making a copy of all the resources but leave out that AuthService we want to drop.

The small amount of extra time it takes to iterate is probably worth saving the memory, but a user with this much annotation config is an edge-case and also the memory saved would be small. I'll commit a change and you can let me know if you think the tradeoff is worth it or a waste.